### PR TITLE
Fix Critical Gradient Flow Issue in DPO Trainer

### PR DESCRIPTION
**Pull request by lizaboiarchuk**
# Fix Critical Gradient Flow Issue in DPO Trainer

## Problem

The DPO trainer had a critical bug preventing gradients from flowing through the model during training, causing model parameters to not update.

### Root Cause

The model's forward pass was computed **outside** the gradient computation context:

**Before (Broken):**
```python
def step(batch, prev_grad, do_update):
    # Policy scores computed OUTSIDE the gradient computation graph
    policy_chosen_scores = get_token_scores(model, chosen, chosen_masks)
    policy_rejected_scores = get_token_scores(model, rejected, rejected_masks)

    policy_chosen_score = compute_score(policy_chosen_scores, chosen_masks, loss_type)
    policy_rejected_score = compute_score(policy_rejected_scores, rejected_masks, loss_type)

    # Gradient computation receives pre-computed scores - no gradient path!
    (lvalue, reward, toks, metrics), grad = loss_value_and_grad(
        policy_chosen_score,
        policy_rejected_score,
        reference_chosen_score,
        reference_rejected_score,
        ...
    )
```

Since `get_token_scores` happened *before* `loss_value_and_grad`, automatic differentiation couldn't track the computational graph through the model.

**Result:**
- ❌ Gradients couldn't backpropagate through the model
- ❌ Model parameters weren't updated
- ❌ Training was ineffective

## Solution

Move the model's forward pass **inside** the `loss_wrapper` so MLX's automatic differentiation can track the full computational graph:

**After (Fixed):**
```python
def loss_wrapper(chosen, rejected, chosen_masks, rejected_masks):
    # Compute policy scores INSIDE the wrapper - gradients can flow!
    policy_chosen_scores = get_token_scores(model, chosen, chosen_masks)
    policy_rejected_scores = get_token_scores(model, rejected, rejected_masks)

    policy_chosen_score = compute_score(policy_chosen_scores, chosen_masks, loss_type)
    policy_rejected_score = compute_score(policy_rejected_scores, rejected_masks, loss_type)

    # Reference model scores with stop_gradient
    if ref_model is None:
        reference_chosen_score = mx.zeros_like(policy_chosen_score)
        reference_rejected_score = mx.zeros_like(policy_rejected_score)
    else:
        ref_chosen_scores = mx.stop_gradient(get_token_scores(ref_model, chosen, chosen_masks))
        ref_rejected_scores = mx.stop_gradient(get_token_scores(ref_model, rejected, rejected_masks))
        reference_chosen_score = compute_score(ref_chosen_scores, chosen_masks, loss_type)
        reference_rejected_score = compute_score(ref_rejected_scores, rejected_masks, loss_type)

    return loss_fn(...)

loss_value_and_grad = nn.value_and_grad(model, loss_wrapper)

@partial(mx.compile, inputs=state, outputs=state)
def step(batch, prev_grad, do_update):
    # Wrapper receives raw inputs, not pre-computed scores
    (lvalue, reward, toks, metrics), grad = loss_value_and_grad(
        chosen, rejected, chosen_masks, rejected_masks
    )

    # Gradient accumulation (unchanged from original)
    if prev_grad is not None:
        grad = mx.tree_map(lambda x, y: x + y, grad, prev_grad)

    if do_update:
        grad = average_gradients(grad)
        if args.gradient_accumulation_steps > 1:
            grad = mx.tree_map(lambda x: x / args.gradient_accumulation_steps, grad)
        optimizer.update(model, grad)
        mx.eval(state)
        grad = None

    return lvalue, reward, toks, metrics, grad
```

### 1. Restructured `loss_wrapper` function (lines 278-318)
- **Changed signature**: Now receives raw batch inputs (`chosen`, `rejected`, `chosen_masks`, `rejected_masks`) instead of pre-computed scores
- **Moved model forward pass inside**: `get_token_scores` and `compute_score` now execute within the gradient computation context

### 2. Updated `step` function (lines 322-342)
- **Changed inputs to `loss_value_and_grad`**: Now passes raw batch data instead of pre-computed scores

### 3. Removed unused parameter (line 252-261)
- **Removed `loss_type` parameter** from `train_dpo` signature (now using `args.loss_type`)

## Impact

- ✅ **DPO training works correctly** - gradients flow through the full model
- ✅ **Model parameters update properly** - backpropagation through complete computational graph

## Testing

To verify the fix:
1. Run DPO training and confirm:
   - Loss decreases over iterations
   - Model parameters change (verify with checksums)
   - Training metrics improve
2. Verify gradient flow: `grad` has non-zero values after `loss_value_and_grad`

## Files Changed

- `mlx_lm_lora/trainer/dpo_trainer.py`: Fixed gradient flow by moving model forward pass inside gradient computation context

### CODE REVIEW

**Comment by lizaboiarchuk** [comment]
Hi @Goekdeniz-Guelmez!
Thank you a lot for your work on MLX LoRA pipelines🔥
We're actively using your package at MacPaw to train our models:) 
We have discovered a gradient flow issue in the DPO trainer.
Could you please review when you have time?

Best regards,
Yelyzaveta
